### PR TITLE
build: Bump Boost version to 1.79

### PR DESCRIPTION
**Pull request by adamemerson**
This fixes the major blocker with enabling C++20 and also allows us to use C++20 coroutines in asynchronous code.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
there are some developer docs about boost in https://docs.ceph.com/en/latest/dev/continuous-integration/#build-dependencies

**Comment by tchaikov** [comment]
Hi Adam, as Casey pointed out. in addition to precompiling boost, we also need to upload the tarball to `https://download.ceph.com/qa/`.

File: install-deps.sh [hunk 1]
```diff
@@ -168,7 +168,7 @@ function install_pkg_on_ubuntu {
 
 function install_boost_on_ubuntu {
     in_jenkins && echo "CI_DEBUG: Running install_boost_on_ubuntu() in install-deps.sh"
-    local ver=1.75
+    local ver=1.79
```
Comment (tchaikov) [inline install-deps.sh:171]: this does not work if we don't have precompiled boost packages uploaded to chacra.

Comment (djgalloway) [reply install-deps.sh:171]: See also https://github.com/ceph/ceph/commit/89121189ac9604913a2a17dc1a224e64cfdde3a1

I will upload the tarball.

Comment (djgalloway) [reply install-deps.sh:171]: Oh, sorry, I pushed the tarball to download.ceph.com.  I'll look into the other bit.

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by adamemerson** [comment]
@djgalloway Could you please build/upload debs for Boost? It think that's the only thing blocking this PR (and I'm pretty sure I don't have access) and this PR is the only thing blocking C++20.

**Comment by djgalloway** [comment]
jenkins retest this please

**Comment by djgalloway** [comment]
@adamemerson I built packages.  This was my first time attempting it so we may still hit some snags.

https://shaman.ceph.com/repos/libboost/master/892ab89e76b91b505ffbf083f6fb7f2a666d4132/default/256966/

**Comment by adamemerson** [comment]
> @adamemerson I built packages. This was my first time attempting it so we may still hit some snags.
> 
> https://shaman.ceph.com/repos/libboost/master/892ab89e76b91b505ffbf083f6fb7f2a666d4132/default/256966/

Thank you very much!

**Comment by djgalloway** [comment]
https://github.com/ceph/ceph/blob/bf63fa67201537b2604a0881c5273d302600c68e/install-deps.sh#L186 needs to change to 892ab89e76b91b505ffbf083f6fb7f2a666d4132

**Comment by adamemerson** [comment]
> 892ab89e76b91b505ffbf083f6fb7f2a666d4132

Done! Thank you.

**Comment by djgalloway** [comment]
Looks like https://github.com/ceph/ceph/blob/main/win32_deps_build.sh#L27-L28 also need to be bumped

**Comment by adamemerson** [comment]
> Looks like https://github.com/ceph/ceph/blob/main/win32_deps_build.sh#L27-L28 also need to be bumped

Done, thank you.

**Comment by tchaikov** [comment]
```
../build.deps/mingw/boost/include/boost/align/aligned_allocator.hpp: In member function 'void boost::alignment::aligned_allocator<T, Alignment>::deallocate(boost::alignment::aligned_allocator<T, Alignment>::pointer, boost::alignment::aligned_allocator<T, Alignment>::size_type)':
../src/include/compat.h:352:27: error: '_aligned_free' is not a member of 'boost::alignment'; did you mean 'aligned_free'?
  352 | #define aligned_free(ptr) _aligned_free(ptr)
      |                           ^~~~~~~~~~~~~
```

turns out `aligned_free` is a member of `boost::alignment`, so using a hackish `#define` to override the default behavior of `aligned_free()` fails the build.

i am proposing a fix to address this. see #47212

**Comment by tchaikov** [comment]
jenkins test windows

**Comment by tchaikov** [comment]
jenkins test make check

**Review by tchaikov — APPROVED** [review]

**Comment by djgalloway** [comment]
I'm okay with this as long as `make check` passes.  Please don't merge until Friday morning at the earliest, however.  I want to be around for fallout if necessary.

And thank you so much @tchaikov for your help.  I absolutely would not have figured out that compilation issue you fixed in https://github.com/ceph/ceph/pull/47212.  We miss you!

**Comment by tchaikov** [comment]
```
In file included from ../src/dokan/options.cc:14:
../src/dokan/ceph_dokan.h:16:15: error: conflicting declaration 'typedef DWORD NTSTATUS'
   16 | typedef DWORD NTSTATUS;
      |               ^~~~~~~~
In file included from ../build.deps/mingw/boost/include/boost/asio/impl/connect_pipe.ipp:29,
                 from ../build.deps/mingw/boost/include/boost/asio/connect_pipe.hpp:79,
                 from ../build.deps/mingw/boost/include/boost/asio.hpp:64,
                 from ../src/include/win32/winsock_wrapper.h:20,
                 from <command-line>:
/usr/share/mingw-w64/include/bcrypt.h:27:16: note: previous declaration as 'typedef LONG NTSTATUS'
   27 |   typedef LONG NTSTATUS,*PNTSTATUS;
      |                ^~~~~~~~
```

**Comment by tchaikov** [comment]
i am proposing #47217 to address this issue. the proposed PR includes the commits in this PR to verify my change.

**Review by tchaikov — CHANGES_REQUESTED** [review]
@adamemerson could you update `make-dist` accordingly? otherwise we'd have FTBFS like
```
-- BUILDING Boost Libraries at j 48
CMake Error at cmake/modules/BuildBoost.cmake:31 (message):
  Boost v1.75.0 in
  /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-13730-g5623bc3c/rpm/el8/BUILD/ceph-17.0.0-13730-g5623bc3c/src/boost
  is not new enough.  Please either "rm -rf
  /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-13730-g5623bc3c/rpm/el8/BUILD/ceph-17.0.0-13730-g5623bc3c/src/boost"
  so I can download Boost v1.79 for you, or make sure
  /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-13730-g5623bc3c/rpm/el8/BUILD/ceph-17.0.0-13730-g5623bc3c/src/boost
  contains a copy of Boost v1.79.
Call Stack (most recent call first):
  cmake/modules/BuildBoost.cmake:147 (check_boost_version)
  cmake/modules/BuildBoost.cmake:251 (do_build_boost)
  CMakeLists.txt:647 (build_boost)
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/64111//consoleFull

**Comment by adamemerson** [comment]
> @adamemerson could you update `make-dist` accordingly? otherwise we'd have FTBFS like

Done!

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test windows

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/pr-46554-kefu/0c4e48e672b8894c273bcd779235bcdcfd9f77f5/

**Comment by djgalloway** [comment]
@tchaikov The RGW suite is seeing valgrind issues that may be related to boost.  Not due to 1.79 explicitly but we were wondering if you could take a look at https://tracker.ceph.com/issues/56500#note-2.  Do you understand what's going on there?

**Comment by tchaikov** [comment]
@djgalloway will take a look tomorrow. just saw your comment right before i am about to call it a day.

**Comment by cbodley** [comment]
essentially, the rgw/verify suite needs a way to run its valgrind jobs against builds (both ceph and boost) with the valgrind flag enabled

would it be possible to:
* build these `ceph-libboost-` packages with the [valgrind=on](https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/stack/valgrind.html) flag
* set `WITH_BOOST_VALGRIND=ON` for all ceph builds that link to these packages

where would this logic need to go? i see `install_boost_on_ubuntu()` in https://github.com/ceph/ceph/blob/main/install-deps.sh#L169, but nothing for centos

**Comment by cbodley** [comment]
cc @yuvalif

**Comment by tchaikov** [comment]
@cbodley i see you scheduled [couple rgw:verify tests](https://pulpito.ceph.com/cbodley-2022-07-26_13:09:16-rgw:verify-pr-46554-kefu-distro-default-smithi/) with [pr-46554-kefu](https://shaman.ceph.com/builds/ceph/pr-46554-kefu/0c4e48e672b8894c273bcd779235bcdcfd9f77f5/default/311571/). they all completed without reporting issues. but according to the [jenkins build log](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=focal,DIST=focal,MACHINE_SIZE=gigantic/64205//consoleFull)

```
CEPH_EXTRA_CMAKE_ARGS=-DALLOCATOR=tcmalloc
...
	cd obj-x86_64-linux-gnu && cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_AUTOGEN_VERBOSE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DWITH_JAEGER=ON -DWITH_SYSTEM_UTF8PROC=ON -DWITH_OCF=ON -DWITH_LTTNG=ON -DWITH_MGR_DASHBOARD_FRONTEND=OFF -DWITH_PYTHON3=3 -DWITH_CEPHFS_JAVA=ON -DWITH_CEPHFS_SHELL=ON -DWITH_SYSTEMD=ON -DCEPH_SYSTEMD_ENV_DIR=/etc/default -DWITH_GRAFANA=ON -DWITH_RBD_RWL=ON -DWITH_RBD_SSD_CACHE=ON -DCMAKE_INSTALL_LIBDIR=/usr/lib -DCMAKE_INSTALL_LIBEXECDIR=/usr/libexec -DCMAKE_INSTALL_SYSCONFDIR=/etc -DBOOST_J=48 -DALLOCATOR=tcmalloc ..
...
-- BUILDING Boost Libraries at j 48
-- boost (1.79.0 >= 1.79) already in /build/ceph-17.0.0-13785-g0c4e48e6/src/boost
...
cd /build/ceph-17.0.0-13785-g0c4e48e6/src/boost && CC=/usr/bin/cc CXX=/usr/bin/c++ ./b2 -j48 -d0 --user-config=/build/ceph-17.0.0-13785-g0c4e48e6/obj-x86_64-linux-gnu/user-config.jam toolset=gcc python=3.8 headers stage variant=release threading=multi link=static address-model=64 "cxxflags=-fPIC -w"
```
the packages were not built with [valgrind=on](https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/stack/valgrind.html) or `WITH_BOOST_VALGRIND=ON`. so i am afraid that your analysis cannot explain the completed runs on ubuntu? 

the `ceph-libbooost-` packages are not involved in qa runs. they are only used when performing "make check" tests performed by jenkins.

if `WITH_BOOST_VALGRIND=ON` is proved to the cure. we need add it at https://github.com/ceph/ceph-build/blob/4aa26880bb7917fa65b5ed11242b4f59aae91a77/scripts/build_utils.sh#L831

**Comment by yuvalif** [comment]
> cc @yuvalif

i tried with the change. did not help. e.g.
http://pulpito.front.sepia.ceph.com/yuvalif-2022-07-28_04:22:02-rgw:verify-wip-yuval-fix-lua-post-ctx-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/yuvalif-2022-07-28_04:20:01-rgw:verify-wip-yuval-lua-filter-distro-default-smithi/

**Comment by cbodley** [comment]
> so i am afraid that your analysis cannot explain the completed runs on ubuntu?

thanks @tchaikov. i didn't expect `WITH_BOOST_VALGRIND` to be on there, i just wanted to test against ubuntu with recent builds, and yours could help rule out the boost version also. that first passing run didn't include any `valgrind` jobs at all, so i had to reschedule in https://pulpito.ceph.com/cbodley-2022-07-26_14:01:08-rgw:verify-pr-46554-kefu-distro-default-smithi/. that result confirmed that the valgrind issues are effecting ubuntu also - i reported on this in https://tracker.ceph.com/issues/56500#note-8

**Comment by djgalloway** [comment]
@tchaikov ISTR you asking to build boost for arm64 too.  Done.  https://shaman.ceph.com/repos/libboost/master/892ab89e76b91b505ffbf083f6fb7f2a666d4132/default/257913/

**Comment by tchaikov** [comment]
> @tchaikov ISTR you asking to build boost for arm64 too. Done. https://shaman.ceph.com/repos/libboost/master/892ab89e76b91b505ffbf083f6fb7f2a666d4132/default/257913/

@djgalloway yes, thank you! to be on par with amd64 nodes. Thank you!

**Pull request merged**